### DISCUSSION TITLE
Why does the code use `if-else` instead of `switch-case`?

### DISCUSSION THREAD

**Discussion by donno2048** [microsoft/terminal#13186] [General]
I guess
https://github.com/microsoft/terminal/blob/11b810e4036b669e5f223cce447f76d56044be0f/src/host/CommandNumberPopup.cpp#L135-L148
https://github.com/microsoft/terminal/blob/11b810e4036b669e5f223cce447f76d56044be0f/src/host/ConsoleArguments.cpp#L435-L470
https://github.com/microsoft/terminal/blob/11b810e4036b669e5f223cce447f76d56044be0f/src/host/ConsoleArguments.cpp#L478-L533
https://github.com/microsoft/terminal/blob/11b810e4036b669e5f223cce447f76d56044be0f/src/inc/til/string.h#L16-L23
make some sense but
https://github.com/microsoft/terminal/blob/11b810e4036b669e5f223cce447f76d56044be0f/src/tools/echokey/main.cpp#L56-L173
https://github.com/microsoft/terminal/blob/11b810e4036b669e5f223cce447f76d56044be0f/src/propsheet/ColorControl.cpp#L7-L26
?

**Comment by zadjii-msft** [comment]
I don't think there's necessarily a reason. I suppose I write a lot of python in my free time, so I write the if/else's instinctively, but those samples I think predate my tenure on the project 😅 Any reason you're asking?

**Reply by donno2048** [reply]
It's just that a switch case is more efficient so why not use it?

**Reply by DHowett** [reply]
A few reasons:

1. Most compilers these days are capable of optimizing long `if` chains to the same machine code as a `switch`-`case`, which makes the conversion unnecessary
2. We do not have unlimited time to go refactor every bit of old code, and it works as-is
3. Even if automatic optimization were not possible, the snippets you've identified are not in hot paths and do not need to be more efficient than they are.

Some of the examples you listed include range checks (the ones in til/string and the one in ColorControl, primarily) which would only suffer from conversion to `switch`-`case`. As an example:

```c++
COLORREF GetColorForItem(const int itemId) 
{
    switch (itemId)
    {
    case IDD_TERMINAL_FGCOLOR:
        return g_fakeForegroundColor; 
    case IDD_TERMINAL_BGCOLOR:
        return g_fakeBackgroundColor; 
    case IDD_TERMINAL_CURSOR_COLOR:
        return g_fakeCursorColor; 
    default:
        if (itemId >= IDD_COLOR_1 && itemId <= IDD_COLOR_16) 
        { 
            return AttrToRGB(itemId - IDD_COLOR_1); 
        } 
        break;
    }
    return RGB(0xff, 0x0, 0xff); 
}
```

Given the range, `default` cannot be comprehensive or exhaustive... so we've traded some clarity away to gain the illusion of convenience.

**Comment by MHKarami97** [comment]
To sum it up, if number of conditions is more than 5 or so, prefer SWITCH over IF, otherwise use whatever looks better.

more info : https://stackoverflow.com/a/395965/10233323
